File: integrations/sreon.mjs
import { spawn } from "node:child_process";
import { createInterface } from "node:readline";

export class Sreon {
  constructor(executable) {
    this.process = spawn(executable, [], { stdio: ["pipe", "pipe", "inherit"], windowsHide: true });
    this.sequence = 0;
    this.pending = new Map();
    this.failure = null;
    createInterface({ input: this.process.stdout }).on("line", (line) => {
      let message;
      try { message = JSON.parse(line); } catch { this.fail(new Error("Invalid Sreon response")); return; }
      const callback = this.pending.get(message.id);
      if (!callback) { this.fail(new Error(message.error?.message || "Unexpected Sreon response")); return; }
      this.pending.delete(message.id);
      if (message.error) callback.reject(new Error(message.error.message));
      else callback.resolve(message.result);
    });
    this.process.on("error", (error) => this.fail(error));
    this.process.on("exit", () => this.fail(new Error("Sreon API stopped")));
    this.process.stdin.on("error", (error) => this.fail(error));
  }

  fail(error) {
    this.failure = error;
    for (const callback of this.pending.values()) callback.reject(error);
    this.pending.clear();
  }

  search(q, category = "web", cursor = null) {
    if (this.failure) return Promise.reject(this.failure);
    if (this.pending.size >= 32) return Promise.reject(new Error("Wait for pending searches to finish"));
    const id = ++this.sequence;
    const line = JSON.stringify({ id, method: "search", params: { q, category, cursor } }) + "\n";
    if (Buffer.byteLength(line) > 16384) return Promise.reject(new Error("Request exceeds 16 KiB"));
    return new Promise((resolve, reject) => {
      this.pending.set(id, { resolve, reject });
      this.process.stdin.write(line);
    });
  }

  close() {
    this.process.stdin.end();
    this.process.kill();
  }
}

File: integrations/sreon.py
import json
import subprocess
import threading


class Sreon:
    def __init__(self, executable):
        self.process = subprocess.Popen([str(executable)], stdin=subprocess.PIPE, stdout=subprocess.PIPE, text=True, encoding="utf-8", bufsize=1)
        self.lock = threading.Lock()
        self.sequence = 0

    def search(self, query, category="web", cursor=None):
        with self.lock:
            self.sequence += 1
            request = {"id": self.sequence, "method": "search", "params": {"q": query, "category": category, "cursor": cursor}}
            self.process.stdin.write(json.dumps(request) + "\n")
            self.process.stdin.flush()
            line = self.process.stdout.readline()
            if not line:
                raise RuntimeError("Sreon API stopped")
            response = json.loads(line)
            if "error" in response:
                raise RuntimeError(response["error"]["message"])
            if response.get("id") != self.sequence:
                raise RuntimeError("Unexpected Sreon response")
            return response["result"]

    def close(self):
        self.process.stdin.close()
        try:
            self.process.wait(timeout=20)
        except subprocess.TimeoutExpired:
            self.process.kill()
            self.process.wait()
        self.process.stdout.close()

    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.close()
